llvm/lib/CodeGen/AsmPrinter/DwarfDebug.cpp

```diff
@@ -3668,3 +3668,19 @@ bool DwarfDebug::alwaysUseRanges(const DwarfCompileUnit &CU) const {
     return true;
   return false;
 }
+
+void DwarfDebug::beginCodeAlignment(const MachineBasicBlock &MBB) {
+  auto *SP = MBB.getParent()->getFunction().getSubprogram();
+
+  bool NoDebug =
+      !SP || SP->getUnit()->getEmissionKind() == DICompileUnit::NoDebug;
+
+  if (NoDebug)
+    return;
+
+  auto PrevLoc = Asm->OutStreamer->getContext().getCurrentDwarfLoc();
+  Asm->OutStreamer->emitDwarfLocDirective(
```

Comment:
I think you're right. We control whether or not we call `beginCodeAlignment` in [this](https://github.com/llvm/llvm-project/pull/99496/files#diff-17b326b45ef392288420bed274616afa7df81b27576c96723b3c25f5198dc398R3969) `if` block but if other users call `beginCodeAlignment` unconditionally they would end up with a line 0 `.loc` immediately followed by the correct `.loc`.

The function is named `beginCodeAlignment` so one may argue you should only call it when emitting alignment nops but on the other hand we can easily move the [`if` block](https://github.com/llvm/llvm-project/pull/99496/files#diff-17b326b45ef392288420bed274616afa7df81b27576c96723b3c25f5198dc398R3969) into `beginCodeAlignment` so I'll do that.

Reply:
Looks like it now skips the line-0 for a block that has 1-byte alignment, but not if the block happens to be aligned.

Reply:
I see what you mean.

```
0000000000000000 <f>:
       ; ...
       3: 7e 2b                         jle     0x30 <f+0x30>
       5: 66 2e 0f 1f 84 00 00 00 00 00 nopw    %cs:(%rax,%rax)
       f: 90                            nop
      10: e8 00 00 00 00                callq   0x15 <f+0x15>
      15: e8 00 00 00 00                callq   0x1a <f+0x1a>
      1a: e8 00 00 00 00                callq   0x1f <f+0x1f>
      1f: e8 00 00 00 00                callq   0x24 <f+0x24>
      24: e8 00 00 00 00                callq   0x29 <f+0x29>
      29: e8 00 00 00 00                callq   0x2e <f+0x2e>
      2e: eb e0                         jmp     0x10 <f+0x10>

      30: e8 00 00 00 00                callq   0x35 <f+0x35> ; BB starting here is a loop but has no nops
      35: eb f9                         jmp     0x30 <f+0x30>
```

We will end up with a line 0 record for address 0x30 immediately followed by a record for the correct line number.

The `MCAlignFragment` type only allows for the possibility of nops being emitted, it does not guarantee that they'll be emitted. The addresses aren't known till after the assembler computes layout which happens much later than here. I can't think of a way to handle this case as things currently stand though there may be other debug-info related mechanisms to deal with this that I'm not aware of, what do you suggest?

Reply:
Hmmm yes at the AsmPrinter level we don't know whether we'll need any nops.  That suggests you want to figure this out down in MC when you _do_ know whether you will be emitting any nops. Or, possibly, when encoding the line table, look to see whether there's a line-0 immediately followed by a non-0 at the same offset, and suppress the line-0. Offhand I don't know where those kinds of things happen, other than "somewhere in MC," sorry.

Reply:
> That suggests you want to figure this out down in MC when you do know whether you will be emitting any nops.

That's going to be difficult because the line table has already been encoded by that time. In theory one could evaluate the line table to find the `MCDwarfLineAddrFragment` associated with the address that the nops are being emitted at but that doesn't seem sustainable.

> Or, possibly, when encoding the line table, look to see whether there's a line-0 immediately followed by a non-0 at the same offset, and suppress the line-0.

Dwarf line entries are encoded in a `MCDwarfLineAddrFragment` which contains an expression and fixups to compute the delta between the current symbol and the last symbol encoded with fixups and then eventually end up as a link-time relocation so the absolute address isn't known until that time.

This seems like a design problem in the MC layer, there isn't a way to correlate a fragment with it's DWARF-related counterpart. This is probably partly due to the fact that fragments don't normally need to change once they've been created. I don't think this is solvable within the scope of this patch.

Reply:
> Dwarf line entries are encoded in a MCDwarfLineAddrFragment which contains an expression and fixups to compute the delta between the current symbol and the last symbol encoded with fixups and then eventually end up as a link-time relocation so the absolute address isn't known until that time.

That doesn't sound right. There should be only one link-time relocation per sequence (usually, one per function). Line-table entries are encoded with a line-adjustment and an instruction-offset delta from the previous entry. It should be feasible to detect a zero instruction delta.

That said, it is a broader problem as I've seen these zero deltas in other situations as well. This patch makes it more prevalent but isn't introducing a new problem. At the AsmPrinter level, you have to assume that there _could_ be nops introduced, and attaching line-0 to those nops is one reasonable solution.

Have you considered instead using the .loc of the first instruction of the new block? That is, emit the new block's .loc _before_ emitting the .align directive. This is something that occurred to me just now, I haven't looked at the code to see whether that's feasible.

Reply:
Oh, never mind. The line-table heuristic will put is_stmt on the nop instead of at the top of the block, and that will probably cause other problems. Go ahead with the line-0 idea.